iroh-util/src/lock.rs

```diff
@@ -1,79 +1,304 @@
+                ProcessStatus::Dead => Ok(false),
+                ProcessStatus::Wakekill => Ok(false),
+                ProcessStatus::Parked => Ok(false),
+                ProcessStatus::LockBlocked => Ok(false),
+                ProcessStatus::Unknown(_s) => Ok(false),
+            }
+        }
+        None => Err(anyhow!("couldn't find system process with id {}", pid)),
+    }
+}
+
+/// Report Process ID stored in a lock file
+pub fn read_lock_pid(prog_name: &str) -> Result<Pid, LockError> {
+    let path = crate::iroh_data_path(&format!("{}.lock", prog_name))
+        .map_err(|e| LockError::Uncategorized { source: e })?;
+    read_lock(&path)
+}
+
+fn read_lock(path: &PathBuf) -> Result<Pid, LockError> {
+    let mut file = File::open(&path).map_err(|e| match e.kind() {
+        ErrorKind::NotFound => LockError::NoLock(path.clone()),
+        e => LockError::Uncategorized {
+            source: anyhow!("{}", e),
+        },
+    })?;
+    let mut pid = String::new();
+    file.read_to_string(&mut pid)
+        .map_err(|_| LockError::CorruptLock(path.clone()))?;
+    let pid = pid
+        .parse::<i32>()
+        .map_err(|_| LockError::CorruptLock(path.clone()))?;
+    Ok(Pid::from(pid))
+}
+
+/// LockError is the set of known program lock errors
+#[derive(Error, Debug)]
+pub enum LockError {
+    // lock present when one is not expected
+    #[error("Locked")]
+    Locked(PathBuf),
+    #[error("No lock file at {0}")]
+    NoLock(PathBuf),
+    /// Failure to parse contents of lock file
+    #[error("Corrupt lock file contents at {0}")]
+    CorruptLock(PathBuf),
+    #[error("Cannot determine status of process holding this lock")]
+    ZombieLock(PathBuf),
+    // location for lock no bueno
+    #[error("invalid path for lock file: {source}")]
+    InvalidPath {
+        #[source]
+        source: anyhow::Error,
+    },
+    /// catchall error type
+    #[error("{source}")]
+    Uncategorized {
+        #[from]
+        source: anyhow::Error,
+    },
+}
+
 #[cfg(all(test, unix))]
 mod test {
     use super::*;
 
     fn create_test_lock(name: &str) -> ProgramLock {
         ProgramLock {
-            path: PathBuf::new().join(name),
+            path: PathBuf::from(name),
             lock: None,
         }
     }
 
+    #[test]
+    fn test_corrupt_lock() {
+        let path = PathBuf::from("lock.lock");
+        let mut f = File::create(&path).unwrap();
+        write!(f, "oh noes, not a lock file").unwrap();
+        let e = read_lock(&path).err().unwrap();
+        match e {
+            LockError::CorruptLock(_) => (),
+            _e => {
+                panic!("expected CorruptLock")
+            }
+        }
+    }
+
     #[test]
     fn test_locks() {
         use nix::unistd::{fork, ForkResult::*};
         use std::io::{Read, Write};
         use std::time::Duration;
 
-        // Start we no lock file.
+        // Start with no lock file.
         let _ = std::fs::remove_file("test1.lock");
```

Comment:
here as well, please use a temp directory

Reply:
this is more complicated than it seems. I'll make an issue & address in a follow up